common/base/include/claragenomics/utils/device_buffer.hpp

```diff
@@ -10,20 +10,200 @@
+    /// Buffer to move from (rhs) is left in an empty state (size = capacity = 0), with the original stream and allocator.
+    /// \param rhs The buffer to move.
+    /// \return refrence to this buffer.
+    buffer& operator=(buffer&& rhs)
+    {
+        data_      = std::exchange(rhs.data_, nullptr);
+        size_      = std::exchange(rhs.size_, 0);
+        stream_    = rhs.stream_;
+        allocator_ = rhs.allocator_;
+        return *this;
+    }
+
+    /// \brief This destructor releases the buffer allocation, returning it to the allocator.
+    ~buffer()
+    {
+        if (nullptr != data_)
+        {
+            allocator_.deallocate(data_, size_);
+        }
+    }
+
+    /// \brief This method returns a pointer to this buffer's first element.
+    /// \return A pointer to the first element of this buffer.
+    value_type* data() { return data_; }
+
+    /// \brief This method returns a const_pointer to this buffer's first element.
+    /// \return A const pointer to the first element of this buffer.
+    const value_type* data() const { return data_; }
+
+    /// \brief This method returns the number of elements in this buffer.
+    size_type size() const { return size_; }
+
+    /// \brief This method returns an iterator pointing to the beginning of this buffer.
+    iterator begin() { return data_; }
+
+    /// \brief This method returns an const_iterator pointing to the beginning of this buffer.
+    const_iterator begin() const { return data_; }
+
+    /// \brief This method returns a iterator pointing to one element past the last of this buffer.
+    /// \return begin() + size().
+    iterator end() { return data_ + size_; }
+
+    /// \brief This method returns a const_iterator pointing to one element past the last of this buffer.
+    /// \return begin() + size().
+    const_iterator end() const { return data_ + size_; }
+
+    /// \brief This method returns the associated stream.
+    /// \return associated stream.
+    cudaStream_t get_stream() const { return stream_; }
+
+    /// \brief This method returns the allocator used to allocate memory for this buffer.
+    /// \return allocator.
+    allocator_type get_allocator() const { return allocator_; }
+
+    /// \brief This method releases the memory and resizes the buffer to 0.
+    void clear_and_free()
+    {
+        if (size_ > 0)
+        {
+            allocator_.deallocate(data_, size_);
+            data_ = nullptr;
+            size_ = 0;
+        }
+    }
+
+    /// \brief This method resizes buffer and discards old data
+    ///
+    /// If old data is still needed it is recommended to crete a new buffer and manually copy the data from the old buffer
+    ///
+    /// \param new_size Number of elements
+    void clear_and_resize(const size_type new_size)
+    {
+        assert(new_size >= 0);
+
+        if (new_size == size_)
+        {
+            return;
+        }
+
+        clear_and_free();
+        size_ = new_size;
```

Comment:
Optional: I know we are not really exception safe anyways, but assigning the size after the data, would at least guarantee that the buffer is of size 0 if the allocation fails.